src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarReader.cs

```diff
@@ -0,0 +1,424 @@
+        /// <para>The default value is <see langword="false"/>.</para></param>
+        /// <returns>A <see cref="TarEntry"/> instance if a valid entry was found, or <see langword="null"/> if the end of the archive has been reached.</returns>
+        /// <exception cref="FormatException"><para>The archive is malformed.</para>
+        /// <para>-or-</para>
+        /// <para>The archive contains entries in different formats.</para>
+        /// <para>-or-</para>
+        /// <para>More than one Global Extended Attributes Entry was found in the current <see cref="TarFormat.Pax"/> archive.</para>
+        /// <para>-or-</para>
+        /// <para>Two or more Extended Attributes entries were found consecutively in the current <see cref="TarFormat.Pax"/> archive.</para></exception>
+        /// <exception cref="IOException">An I/O problem ocurred.</exception>
+        public TarEntry? GetNextEntry(bool copyData = false)
+        {
+            Debug.Assert(_archiveStream.CanRead);
+
+            if (_archiveStream.CanSeek && _archiveStream.Length == 0)
+            {
+                // Attempting to get the next entry on an empty tar stream
+                return null;
+            }
+
+            AdvanceDataStreamIfNeeded();
+
+            if (TryGetNextEntryHeader(out TarHeader header, copyData))
+            {
+                if (!_readFirstEntry)
+                {
+                    Debug.Assert(Format == TarFormat.Unknown);
+                    Format = header._format;
+                    _readFirstEntry = true;
+                }
+                else if (header._format != Format)
+                {
+                    throw new FormatException(string.Format(SR.TarEntriesInDifferentFormats, header._format, Format));
+                }
+
+                TarEntry entry = Format switch
+                {
+                    TarFormat.Pax => new PaxTarEntry(header, this),
+                    TarFormat.Gnu => new GnuTarEntry(header, this),
+                    TarFormat.Ustar => new UstarTarEntry(header, this),
+                    TarFormat.V7 or TarFormat.Unknown or _ => new V7TarEntry(header, this),
+                };
+
+                _previouslyReadEntry = entry;
+                PreserveDataStreamForDisposalIfNeeded(entry);
+                return entry;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Asynchronously retrieves the next entry from the archive stream.
+        /// </summary>
+        /// <param name="copyData"><para>Set it to <see langword="true"/> to copy the data of the entry into a new <see cref="MemoryStream"/>. This is helpful when the underlying archive stream is unseekable, and the data needs to be accessed later.</para>
```

Comment:
Why do we need this option rather than the developer just doing it themselves if they need it?